R/tmb-methods.R

```diff
@@ -207,3 +207,111 @@ print.mmrm_tmb <- function(x,
   cat("\n")
   invisible(x)
 }
+
+
+#' Residuals method for mmrm objects
+#'
+#' @describeIn mmrm_tmb_method to obtain residuals - either unscaled or normalized
+#' @param object (`mmrm_tmb`)\cr the fitted MMRM.
+#' @param type (`character`)\cr unscaled ('response'), 'pearson' or 'normalized'. Default is 'response',
+#' and this is the only type available for use with models with a spatial covariance structure.
+#' @param ... \cr not used in this method.
+#' @return Vector of residuals
+#' @importFrom stats residuals
+#' @exportS3Method
+#' @examples
+#' # residuals:
+#' residuals(object, type = "response")
+#' residuals(object, type = "pearson")
+#' residuals(object, type = "normalized")
+#' @references
+#' - \insertRef{galecki2013linear}{mmrm}
+residuals.mmrm_tmb <- function(object, type = c("response", "pearson", "normalized"), ...) {
+  assert_class(object, "mmrm_tmb")
+  assert_character(type)
+  type <- match.arg(type, several.ok = FALSE)
+  resids_unscaled <- component(object, "y_vector") - fitted(object)
+  if (type == "response") {
+    return(unname(resids_unscaled))
+  } else {
+    if (object$formula_parts$is_spatial) {
+      stop("Only raw / response residuals are available for models with spatial covariance structures.")
+    }
+    if (type == "pearson") {
+      h_residuals_pearson(object, resids_unscaled)
+    } else if (type == "normalized") {
+      h_residuals_normalized(object, resids_unscaled)
+    }
+  }
+}
+
+#' Calculate Pearson residuals
+#'
+#' This is used by [residuals.mmrm_tmb()] to calculate Pearson residuals.
+#'
+#' @param object (`mmrm_tmb`)\cr the fitted MMRM.
+#' @param resids_unscaled (`numeric`)\cr the raw/response residuals.
+#'
+#' @return Vector of residuals
+#'
+#' @keywords internal
+h_residuals_pearson <- function(object, resids_unscaled) {
+  visits <- as.numeric(object$tmb_data$full_frame[[object$formula_parts$visit_var]])
+  if (component(object, "n_groups") == 1) {
+    visit_sigmas <- sqrt(diag(object$cov))
+    resids <- resids_unscaled / visit_sigmas[visits] * sqrt(object$tmb_data$weights_vector)
+  } else {
```

Comment:
I think the code here becomes less duplicate/simpler if you first define 
```r
visit_sigmas <- if (component(object, "n_groups") == 1) {
  sqrt(diag(object$cov))
} else {
  # work with grp_visit_sigmas etc.
}
```
and then use that because rest of code is the same for both cases